src/wait-for.js

```diff
@@ -81,15 +81,15 @@ function waitFor(
           jest.advanceTimersByTime(interval)
         })
 
+        // Could have timed-out
+        if (finished) {
```

Comment:
With this factoring, I wonder if we could simplify further by moving `checkCallback` before the `advanceTimers`. That way the loop condition makes sure we exit as soon as we're `finished`. And we can remove another `checkCallback` on L54.

At this point I would remove the "It's really important that checkCallback [...]" comment. We might have fixed the issue along the way. And since I'd rather ship this with the next major, we have a good opportunity to resurface the bug so that we can at least link a repro. It doesn't have to be a Jest test but somehow the reordering fixed "something". That "something" needs to be linked a t least. Doesn't need to be an automated test but it needs to be something. Otherwise we're chasing ghosts.

Reply:
If we move `checkCallback` before the `advanceTimers`, the `waitFor` `callback` will be called twice before we advance the clock, instead of once right now.

Which will break this fix: https://github.com/testing-library/dom-testing-library/pull/1073/files

Reply:
> the waitFor callback will be called twice before we advance the clock, instead of once right now.

but not if we remove the `checkCallback` call earlier from L54, no?

Reply:
There will be a change of behavior in https://github.com/testing-library/dom-testing-library/blob/37ccada9ed26dec9a50bee8df3af4ea0e45d61b9/src/__tests__/fake-timers.js#L43, for a `timeout` smaller than the `interval`, and will instead be rejected with `[Error: Timed out in waitFor.]`